source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -48,66 +50,104 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
   return FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
+FilterHeadersStatus Filter::encodeHeaders(ResponseHeaderMap& headers, bool end_of_stream) {
+  if (stream_closed_) {
+    return FilterHeadersStatus::Continue;
```

Comment:
Should this keep forwarding on a closed management stream or halt? How come we don't reset the stream in `closeStream()`?

Reply:
"closeStream" is supposed to close the stream to the external processor. We just do this by calling "close" on the gRPC stream. Is there something else that we're supposed to to?

As for forwarding on a closed management stream -- yes. The protocol allows the external processor to close the stream at any time, at which point the filter just skips it. The configuration also allows us to set "ignore errors," which means that we treat a failed stream the same as a closed stream. Finally, if there is an error on the stream, then we "sendLocalReply," but the filter still gets invoked after that in some cases so we still want to continue AFAIK.

Reply:
Right, my point was that if you have closed the stream to the management server (e.g. on error) but aren't doing "ignore errors", do you still want to potentially forward traffic?

Reply:
Right now, this is what happens:

decodeHeaders()
error on gRPC stream
filter sets "stream_closed_" and calls sendLocalReply()

at this point, it's been my experience that sendLocalReply will call encodeHeaders, presumably to give us a chance to do more processing. We don't want to do anything else at this point since the processor stream failed so we have no choice IMO but to continue. Is there another option there?

Reply:
FWIW I changed the filter to always call sendLocalReply on encoder_callbacks_, and now the extra call to encodeHeaders doesn't show up. So the only way we should get to encodeHeaders with stream_closed_ set to true is if either the remote processor closed the stream or if we closed it while ignoring an error.